### DISCUSSION TITLE
CustomActivities  dependency injection

### DISCUSSION THREAD

**Discussion by ILAOHE** [elsa-workflows/elsa-core#4128] [Q&A] [answered]
After integrating elsa 3.0 into my app, how do I inject my app's own services into a custom Activity to perform an action? This way of writing is obviously not possible.
<img width="1182" alt="image" src="https://github.com/elsa-workflows/elsa-core/assets/20391853/689dad91-4740-4967-84fd-bfc098458418">

**Comment by MikeDLi** [comment]
There seems to be an issue with custom activity DI in 3.0.0-rc1:

![image](https://github.com/elsa-workflows/elsa-core/assets/99539881/43fe6fbc-82d5-4169-b9be-b46f51fa01be)

and this is the error message:

An unhandled exception has occurred while executing the request.
System.InvalidOperationException: Each parameter in the deserialization constructor on type '[skipped].Activities.TestDI' must bind to an object property or field on deserialization. Each parameter name must match with a property or field on the object. Fields are only considered when 'JsonSerializerOptions.IncludeFields' is enabled. The match can be case-insensitive.
   at System.Text.Json.ThrowHelper.ThrowInvalidOperationException_ConstructorParameterIncompleteBinding(Type parentType)   at System.Text.Json.Serialization.Converters.ObjectWithParameterizedConstructorConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value).

Thanks!

**Comment by sfmskywalker** [comment, marked as answer]
In Elsa 3, activities are simple objects that are not resolved from the service container. Instead, you can simply construct workflow graphs by instantiating activity objects directly.

For example:

```csharp
var workflow = new Sequence
{
   Activities = 
   {
      new WriteLine("Hello world!),
      new SendEmail("everyone@earth.io")
   }
};
```

A consequence of allowing this is that you can no longer use constructor injection. Unless, of course, you resolve your activities from a service container, like this:

```csharp
var workflow = new Sequence
{
   Activities = 
   {
      serviceProvider.GetRequiredService<TestDI>()
   }
};
```

And, if we wanted to support constructor injection also for activities that are part of a JSON document created using the designer, we'd have to update the activity JSON converter to resolve activities from the service provider, too. Which might be a good thing to do, actually.

Without constructor injection (or property injection, for that matter), you can resolve your dependencies from within the activity's ExecuteAsync method, like this:

```csharp
public class TestDI : Activity
{
   public override ValueTask ExecuteAsync(ActivityExecutionContext context)
   {
      var textWriter = context.GetRequiredService<TextWriter>();
   }
}
```

Another approach to construct workflows programmatically while also allowing constructor injection is to build a workflow builder API, where you no longer instantiate activities manually, but instead, rely on the builder API to do so (which internally leverages DI).

**Comment by MikeDLi** [comment]
Thank you so much for your quick response!

**Comment by profi** [comment]
I am exploring EW3. And i think i found a problem. It might be my configuration tho.
I want to access the dbcontext  data and take all the related data  from an activity.

But it doesnt seems to work with Include() command.

I am using separate custom dbcontext.

```
protected override void Execute(ActivityExecutionContext context)
        {

            _context =  context.GetRequiredService<DbContext>();

            var productId= productId.Get(context);

            var product= _context.Products
                 .Include(a => a.Orders)
                 .Include(a => a.Prices)
                .SingleOrDefault(a => a.Id == productId);

            // now products will be single object without its related tables Orders and Prices (1:m)
           //But now if run next line product object will contain related table Orders Data

            var orders= _context.Orders
              .Where(q => q.ProductId== productId)
              .ToList();
          
        }
```
  Out of the activity in a standard way(or repository class method) requesting the data from the database with the same code works as it should.
          
I dont want to make multiple requests to the database. What might be the problem?
